### PR TITLE
UGENE-7072. Separate 'sequence similarity' and 'tree-guided' grouping…

### PR DESCRIPTION
**Pull request by yalgaer**
… modes in MSA editor.

### CODE REVIEW

File: src/corelibs/U2View/src/ov_msa/Export/MSAImageExportTask.cpp [hunk 1]
```diff
@@ -306,16 +306,9 @@ void MSAImageExportController::initSettingsWidget() {
...
-        for (qint64 i = selection.y(); i < selection.height() + selection.y(); i++) {
-            msaSettings.seqIdx.append(i);
-        }
-    } else {
-        MaCollapseModel *model = ui->getCollapseModel();
-        SAFE_POINT(model != NULL, tr("MSA Collapsible Model is NULL"), );
-        for (qint64 i = selection.y(); i < selection.height() + selection.y(); i++) {
-            msaSettings.seqIdx.append(model->getMaRowIndexByViewRowIndex(i));
-        }
+    MaCollapseModel *model = ui->getCollapseModel();
```
Comment (evelinabis) [inline src/corelibs/U2View/src/ov_msa/Export/MSAImageExportTask.cpp:313]: Can `ui` and `model` be `nullptr` here and elsewhere?

File: src/corelibs/U2View/src/ov_msa/MSAEditor.cpp [hunk 2]
```diff
@@ -772,7 +772,7 @@ void MSAEditor::sortSequences(const MultipleAlignment::SortType &sortType, const
     msa->sortRows(sortType, sortOrder, sortRange);
 
     // Disable virtual mode.
```
Comment (evelinabis) [inline src/corelibs/U2View/src/ov_msa/MSAEditor.cpp]: Perhaps comment is out of date

File: src/plugins/GUITestBase/src/GTUtilsMsaEditorSequenceArea.cpp [hunk 3]
```diff
@@ -360,14 +360,14 @@ bool GTUtilsMSAEditorSequenceArea::isCollapsed(GUITestOpStatus &os, QString seqN
...
-    QAbstractButton *collapce = GTAction::button(os, "Enable collapsing");
+    QAbstractButton *toggleSequenceOrderButton = GTAction::button(os, "toggle_sequence_row_order_action");
     bool nameLists = getVisibleNames(os) == getNameList(os);
-    if (nameLists && !collapce->isChecked()) {
+    if (nameLists && !toggleSequenceOrderButton->isChecked()) {
         return false;
-    } else if (!nameLists && collapce->isChecked()) {
+    } else if (!nameLists && toggleSequenceOrderButton->isChecked()) {
         return true;
     } else {
```
Comment (evelinabis) [inline src/plugins/GUITestBase/src/GTUtilsMsaEditorSequenceArea.cpp]: "Else" is not needed

File: src/corelibs/U2View/src/ov_msa/MaEditor.h [hunk 4]
```diff
@@ -69,6 +69,31 @@ class SNPSettings {
...
+     * This mode allows to re-order read-only MA objects.
+     */
+    Sequence,
+
+    /**
+     * Sequences are ordered by some external manager (e.g. tree-view) and can be re-ordered by user (e.g. drag & drop).
+     * The order is not saved to the file and is kept in the view memory only.
+     * This mode allows to re-order read-only MA objects.
+     */
+    Free
```
Comment (evelinabis) [inline src/corelibs/U2View/src/ov_msa/MaEditor.h:94]: Is it possible to distinguish between these cases (tree-view and drag & drop)?

File: src/plugins/GUITestBase/src/tests/common_scenarios/msa_editor/tree/GTTestsMSAEditorTree.cpp [hunk 5]
```diff
@@ -75,5 +78,58 @@ GUI_TEST_CLASS_DEFINITION(test_0001) {
...
+    GTUtilsProjectTreeView::toggleView(os);    // Close project view to make all actions on toolbar available.
+
+    GTUtilsDialog::waitForDialog(os, new BuildTreeDialogFiller(os, sandBoxDir + "msa_editor_tree_test_0002", 0, 0, true));
+    GTWidget::click(os, GTAction::button(os, "Build Tree"));
+    GTUtilsTaskTreeView::waitTaskFinished(os);
+
+    QStringList byTreeSequenceNames1 = GTUtilsMSAEditorSequenceArea::getVisibleNames(os);
+    QAbstractButton *syncModeButton = GTAction::button(os, "sync_msa_action");
+    QAbstractButton *toggleSequenceOrderButton = GTAction::button(os, "toggle_sequence_row_order_action");
+    QAbstractButton *refreshSequenceOrderButton = GTAction::button(os, "refresh_sequence_row_order_action");
```
Comment (evelinabis) [inline src/plugins/GUITestBase/src/tests/common_scenarios/msa_editor/tree/GTTestsMSAEditorTree.cpp:99]: Why isn't pressing this button checked?

File: src/corelibs/U2View/src/ov_msa/Export/MSAImageExportTask.cpp (hunk 1 above)
Comment (yalgaer) [reply src/corelibs/U2View/src/ov_msa/Export/MSAImageExportTask.cpp:313]: `ui` or `model` can't be null in MSAImageExportController or in any other place where it is expected to be passed/stored.

However there is a design issue when some widgets are cached in async tasks without smart pointers, but this is a different story.

File: src/plugins/GUITestBase/src/GTUtilsMsaEditorSequenceArea.cpp (hunk 3 above)
Comment (yalgaer) [reply src/plugins/GUITestBase/src/GTUtilsMsaEditorSequenceArea.cpp]: Fixed

File: src/corelibs/U2View/src/ov_msa/MSAEditor.cpp (hunk 2 above)
Comment (yalgaer) [reply src/corelibs/U2View/src/ov_msa/MSAEditor.cpp]: Fixed

File: src/corelibs/U2View/src/ov_msa/MaEditor.h (hunk 4 above)
Comment (yalgaer) [reply src/corelibs/U2View/src/ov_msa/MaEditor.h:94]: There is no need today to differentiate between any externally managed row ordering modes in the MSA Editor, so it does not matter if this is a tree guided sort or any other kind of sort injected by a 3rd party plugin.
In the Free mode MSA editor should not save changes to the file (the mode is virtual) and should not dynamically recompute sequence content similarity groups.

Once we will have a use case that need the differentiation we will support it.

File: src/plugins/GUITestBase/src/tests/common_scenarios/msa_editor/tree/GTTestsMSAEditorTree.cpp (hunk 5 above)
Comment (yalgaer) [reply src/plugins/GUITestBase/src/tests/common_scenarios/msa_editor/tree/GTTestsMSAEditorTree.cpp:99]: Could you please elaborate what do you mean by "pressing this button checked"?

Comment (evelinabis) [reply src/plugins/GUITestBase/src/tests/common_scenarios/msa_editor/tree/GTTestsMSAEditorTree.cpp:99]: Question is canceled

**Review by evelinabis — APPROVED** [review]

File: src/corelibs/U2View/src/ov_msa/MSAEditorSequenceArea.cpp [hunk 6]
```diff
@@ -227,15 +227,25 @@ void MSAEditorSequenceArea::updateCollapseModel(const MaModificationInfo &modInf
...
+            sl_toggleSequenceRowOrder(false);
+        }
+        return;
     }
 
-    // Create row groups by similarity. Do not modify the alignment.
+    SAFE_POINT(mode == MaEditorRowOrderMode::Sequence, "Unexpected row order mode", );
+
+    // Order and group rows by sequence content.
+    MultipleSequenceAlignmentObject *msaObject = getEditor()->getMaObject();
```
Comment (DmitriiSukhomlinov) [inline src/corelibs/U2View/src/ov_msa/MSAEditorSequenceArea.cpp:248]: `SAFE_POINT(msaObject != nullptr, );`

File: src/corelibs/U2View/src/ov_msa/MaEditorNameList.cpp [hunk 7]
```diff
@@ -190,6 +188,10 @@ void MaEditorNameList::updateScrollBar() {
     connect(nhBar, SIGNAL(valueChanged(int)), SLOT(sl_completeRedraw()));
 }
 
+int MaEditorNameList::getGroupExpanderWidth() const {
+    return ui->isCollapsingOfSingleRowGroupsEnabled() || ui->getCollapseModel()->hasGroupsWithMultipleRows() ? 2 * CROSS_SIZE : 0;
```
Comment (DmitriiSukhomlinov) [inline src/corelibs/U2View/src/ov_msa/MaEditorNameList.cpp:192]: Please, add round brackets around the condition

**Review by DmitriiSukhomlinov — APPROVED** [review]
Minor comments

**Pull request merged**